File: app/src/main/java/personal/abdulmughniaf/androidpractice/TodoView.java
package personal.abdulmughniaf.androidpractice;

import android.content.Context;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.widget.CheckBox;
import android.widget.TextView;

import androidx.constraintlayout.widget.ConstraintLayout;

import personal.abdulmughniaf.androidpractice.models.Todo;

public class TodoView extends ConstraintLayout {
    public CheckBox completeCheckBox;
    public TextView descriptionView;

    public TodoView(Context context) {
        super(context);
    }

    public TodoView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public TodoView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void initView (Todo todo){
        completeCheckBox = findViewById(R.id.completeCheckBox);
        descriptionView = findViewById(R.id.descriptionView);

        completeCheckBox.setChecked(todo.isComplete());
        descriptionView.setText(todo.getDescription());

        if (todo.isComplete()){
            createStrikeThrough();
        }
        setupCheckStateListener();
    }

    private void setupCheckStateListener() {
        completeCheckBox.setOnCheckedChangeListener((button, isChecked) ->{
            if (isChecked){
                createStrikeThrough();
            }
            else {
                removeStrikeThrough();
            }
        });
    }

    private void createStrikeThrough() {
        descriptionView.setPaintFlags(descriptionView.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
    }

    private void removeStrikeThrough(){
        descriptionView.setPaintFlags(descriptionView.getPaintFlags() & ~Paint.STRIKE_THRU_TEXT_FLAG);
    }
}

File: app/src/main/java/personal/abdulmughniaf/androidpractice/view/TodoView.kt
package personal.abdulmughniaf.androidpractice.view

import android.content.Context
import android.graphics.Paint
import android.util.AttributeSet
import androidx.constraintlayout.widget.ConstraintLayout
import kotlinx.android.synthetic.main.view_todo.view.*
import personal.abdulmughniaf.androidpractice.models.Todo

class TodoView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 1
): ConstraintLayout(context, attrs, defStyleAttr) {

    fun initView(todo: Todo, callback: (() -> Unit)? = null) {
        completeCheckBox.isChecked = todo.isComplete
        descriptionView.text = todo.description

        if (todo.isComplete) {
            createStrikeThrough()
        }
        setupCheckStateListener(todo, callback)
    }

    private fun setupCheckStateListener(todo: Todo, callback: (() -> Unit)? = null) {
        completeCheckBox.setOnCheckedChangeListener { _, isChecked ->
            todo.isComplete = isChecked
            callback?.invoke()
            if (isChecked) {
                createStrikeThrough()
            } else {
                removeStrikeThrough()
            }
        }
    }

    private fun createStrikeThrough() {
        descriptionView.apply {
            paintFlags = paintFlags or  Paint.STRIKE_THRU_TEXT_FLAG
        }
    }

    private fun removeStrikeThrough() {
        descriptionView.apply {
            paintFlags = paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
        }
    }
}
